### PR TITLE
New transform for Stanford ETL: move visit starts to start time of first non-visit event

### PR DESCRIPTION
**Pull request by scottfleming**
Added a transform, `move_visit_start_to_first_event_start`, that will systematically assign the start time associated with each visit to be the start time of the first event that (1) is associated with the visit (i.e., shares the same visit ID as the visit event), (2) is a non-visit event, and (3) occurs on the same day as the visit event. If the visit has no non-visit events or all events associated with the visit have the same start time as the visit event (e.g., events with a start time of midnight such as billing codes, assuming visit events also have a midnight start time) then the visit start time remains unchanged. Events that occur on days prior to the visit do not affect the visit start time.

This transform should make it so that visit events appear at the same time as or before any non-visit events associated with that visit when events for a given day are ordered by time. (Note that visit events may appear after non-visit events with the same start time in the patient timeline when serialized unless visits are explicitly prioritized). This is in harmony with the idea that clinicians and models should both have access to information about visit type when making predictions about future codes/events for a patient within a given visit.

### CODE REVIEW

File: src/piton/transforms/stanford.py [hunk 1]
```diff
@@ -16,6 +16,103 @@ def _move_date_to_end(
...
+    a start time of '00:00:00'. Events that occur on days prior to the
+    visit do not affect the visit start time.
+
+    NOTE: This function assumes that all non-visit events with the same
+    start time as the visit event (e.g., events with a start time at midnight
+    such as billing codes, in the case where visit events also have a midnight
+    start time) are moved to day end, otherwise there may be non-visit events
+    with start times before the visit start time.
+
+    The reason for assigning a start time of 11:58:59 PM is that it makes
```
Comment (Lalaland) [inline src/piton/transforms/stanford.py:58]: We operate in minute resolution by default, so this should be 11:58 PM

File: src/piton/transforms/stanford.py [hunk 2]
```diff
@@ -16,6 +16,103 @@ def _move_date_to_end(
...
+    that order. In the OMOP DEID from 12/20/2022 about 10% of visits have
+    a time that is not  '00:00:00'.
+    """
+    first_event_starts: Dict[int, datetime.datetime] = {}
+    visit_starts: Dict[int, datetime.datetime] = {}
+
+    # Find the given start time for each visit
+    for event in patient.events:
+        if event.omop_table == "visit":
+            visit_starts[event.visit_id] = event.start
```
Comment (Lalaland) [inline src/piton/transforms/stanford.py:79]: Please add an assert that event.visit_id is not already in visit_starts. It shouldn't be, but would be good to double check.

File: src/piton/transforms/stanford.py [hunk 3]
```diff
@@ -16,6 +16,103 @@ def _move_date_to_end(
...
+
+    # Find the minimum start time over all non-visit events associated with each visit
+    for event in patient.events:
+        if event.visit_id is not None:
+            # Ignore any events for which event start is before the associated visit start
```
Comment (Lalaland) [inline src/piton/transforms/stanford.py:84]: Make the text match the following if statement.

For example, this should be "Only trigger when the event start is after the visit start"

File: src/piton/transforms/stanford.py [hunk 4]
```diff
@@ -16,6 +16,103 @@ def _move_date_to_end(
...
+            # Also ignore non-visit events starting same time as visit (i.e., at midnight)
+            if event.start > visit_starts[event.visit_id]:
+                if event.visit_id in first_event_starts:
```
Comment (Lalaland) [inline src/piton/transforms/stanford.py:87]: Minor nit, you can avoid this if statement by using Python dict's .get method which lets you supply a default time.

File: src/piton/transforms/stanford.py [hunk 5]
```diff
@@ -16,6 +16,103 @@ def _move_date_to_end(
...
+    for event in patient.events:
+        if event.omop_table == "visit":
+            if event.visit_id in first_event_starts:
+                first_event_time = first_event_starts[event.visit_id]
+                event.start = first_event_time - datetime.timedelta(seconds=1)
+            elif event.start.time() == datetime.time.min:
+                event.start = (
+                    event.start
+                    + datetime.timedelta(days=1)
+                    - datetime.timedelta(seconds=61)
```
Comment (Lalaland) [inline src/piton/transforms/stanford.py:104]: Let's not use seconds here. Keep things in minutes per the default time resolution.

File: src/piton/transforms/stanford.py [hunk 6]
```diff
@@ -16,6 +16,103 @@ def _move_date_to_end(
...
+                    )
+                else:
+                    first_event_starts[event.visit_id] = event.start
+
+    # Assign visit start times to be one second before the first event associated with that visit
+    for event in patient.events:
+        if event.omop_table == "visit":
+            if event.visit_id in first_event_starts:
+                first_event_time = first_event_starts[event.visit_id]
+                event.start = first_event_time - datetime.timedelta(seconds=1)
```
Comment (Lalaland) [inline src/piton/transforms/stanford.py:99]: Likewise, keep this in minutes, not seconds

**Review by Lalaland — CHANGES_REQUESTED** [review]
Looks mostly good, with some minor changes.

One bigger question, I guess we aren't going to apply similar logic to the visit ends? I guess there isn't as much need.

File: src/piton/transforms/stanford.py (hunk 1 above)
Comment (scottfleming) [reply src/piton/transforms/stanford.py:58]: Sounds good - I was back and forth about seconds vs. minutes but I think that's the right call for consistency.

File: src/piton/transforms/stanford.py (hunk 2 above)
Comment (scottfleming) [reply src/piton/transforms/stanford.py:79]: Good catch.

File: src/piton/transforms/stanford.py (hunk 6 above)
Comment (scottfleming) [reply src/piton/transforms/stanford.py:99]: I hear you - my concern here is the case where the first non-visit event has a timestamp of 12:01 AM. In that case taking off one minute from the first event start time would put the visit start back to midnight exactly, at which point moving all midnight events to the end of the day would push the visit start past the first event associated with it.

Comment (scottfleming) [reply src/piton/transforms/stanford.py:99]: I suppose I could just do a special case where we set the new visit start time to be max(first event time - 1 minute, 12:01 AM)

Comment (Lalaland) [reply src/piton/transforms/stanford.py:99]: Yeah, let's just handle the edge case manually to 12:01 AM

**Comment by Lalaland** [comment]
Hi Scotty, thanks for the work.

I'm not really a fan of prioritize_visit_events. Why not simply add an explicit priority column and add that column to all the corresponding sort orders?

Having a sort order being applied in this sort of weird manner seems like it would cause a ton of issues. In particular, I think we want "resort" to always fix things, which is not the case with this transformation.

**Comment by scottfleming** [comment]
Sure yeah we don't have to include the `prioritize_visit_events`, happy to drop it. It was basically just an attempt at a patch to have visits appear first when there are multiple simultaneous events in the meantime. In that case, I probably wouldn't include priority-based sorting as part of this PR and we'd just stick with having visits assigned the same start time as the first event for this one.

**Comment by Lalaland** [comment]
Let's just drop prioritize_visit_events for now. The 2x overhead of two pass visit processing shouldn't be a big deal for now.

And the two pass algorithm is much simpler as well.

**Review by rthapa84 — APPROVED** [review]
These looks good to me.

**Comment by scottfleming** [comment]
> Let's just drop prioritize_visit_events for now. 
@Lalaland done.

> The 2x overhead of two pass visit processing shouldn't be a big deal for now. And the two pass algorithm is much simpler as well.
I'm not sure what you mean by the two-pass algorithm here? Are you suggesting post-processing patient timelines from the ETL?

**Review by scottfleming — COMMENTED** [review]
Changes reviewed and incorporated

**Comment by Lalaland** [comment]
> I'm not sure what you mean by the two-pass algorithm here? Are you suggesting post-processing patient timelines from the ETL?

With the current code, in order to process visits in downstream code (labelers, other transforms, featurization, etc) you have to do two passes over the patient. If we fixed the ordering, it would only take one pass.

**Review by Lalaland — APPROVED** [review]
Looks good! Feel free to merge whenever you are ready.

**Pull request merged**